src/rtc_peer.ts

```diff
@@ -123,6 +170,117 @@ export class RTCPeer extends EventEmitter {
+        if (!av1Codec) {
+            this.logger.logDebug('RTCPeer.handleCodecSupportUpdate: client does not support AV1 codec, returning');
+            return false;
+        }
+
+        const vp8Codec = await RTCPeer.getVideoCodec(CodecMimeType.VP8);
+        if (!vp8Codec) {
+            // Realistically, this should never happen.
+            this.logger.logErr('RTCPeer.handleCodecSupportUpdate: client does not support VP8 codec, returning');
+            return false;
+        }
+
+        // Second, we check AV1 support level of the call. Partial or None levels are treated the same as we don't want to send
+        // multiple encodings at the same time. This means we need full support to send an AV1-encoded track.
+        const av1CallSupport = supportMap[CodecMimeType.AV1] === CodecSupportLevel.Full;
+
+        // Now we check whether we need to make any changes to our encodings for outgoing tracks (senders).
+        // If av1CallSupport is true, we need to ensure that we are sending AV1 (if we are sending any video tracks that is)
+        // Else, if av1CallSupport is false, we need to ensure we are sending video tracks using VP8 and switch codec where necessary.
+        const targetCodec = av1CallSupport ? av1Codec : vp8Codec;
+
+        let needsNegotiation = false;
+        for (const tctx of Object.values(this.trackCtxs)) {
+            if (tctx.sender.track?.kind !== 'video') {
+                // Skip non-video tracks.
+                continue;
+            }
+
+            this.logger.logDebug(`RTCPeer.handleCodecSupportUpdate: av1CallSupport=${av1CallSupport} checking video sender`, tctx);
+
+            const params = tctx.sender.getParameters();
+            const currentCodec = params.codecs[0];
+
+            // Only switch if we're not already sending the track using the target codec.
+            if (currentCodec.mimeType !== targetCodec.mimeType) {
+                this.logger.logDebug(`RTCPeer.handleCodecSupportUpdate: ${targetCodec.mimeType} codec not used for video sender, need to switch encoder to ${targetCodec.mimeType}`, av1CallSupport, currentCodec, tctx);
+
+                // eslint-disable-next-line no-await-in-loop
+                await this.switchCodecForTrack(tctx, targetCodec);
+
+                needsNegotiation = true;
+            }
+        }
+
+        const existingTransceiver = this.pc?.getTransceivers().find((trx) => {
+            if (!trx.receiver) {
+                return false;
+            }
+            return trx.receiver.track && this.mediaMap[trx.mid!]?.mime_type === targetCodec.mimeType;
+        });
+
+        if (existingTransceiver) {
+            this.logger.logDebug(`RTCPeer.handleCodecSupportUpdate: ${targetCodec.mimeType} receiver already exists, need to emit track`, existingTransceiver, existingTransceiver.receiver.track);
+            this.emit('stream', new MediaStream([existingTransceiver.receiver.track]), this.mediaMap[existingTransceiver.mid!]);
```

Comment:
This is why we need to include mime types in media map entries. So that we can re-fire track events and let the upper layer render the right track that's being sent a codec update.